### PR TITLE
Implement the Repository Provider

### PR DESCRIPTION
**Pull request by vaibhavdaren**
This PR implements the  #3995  Repository Provider feature.

### Testing

A hacky run on guest  since we dont have `Plugin Ready`.  Could get the file downloaded and list RPM's 

```
 cmd: curl -L --fail -o /etc/yum.repos.d/docker-ce.repo https://download.docker.com/linux/centos/docker-ce.repo
            err:   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
            err:                                  Dload  Upload   Total   Spent    Left  Speed
            err:

                  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
                100   811  100   811    0     0   9430      0 --:--:-- --:--:-- --:--:--  9430
            cmd: cat /etc/yum.repos.d/docker-ce.repo
            out: [docker-ce-stable]
            out: name=Docker CE Stable - $basearch
            out: baseurl=https://download.docker.com/linux/centos/$releasever/$basearch/stable
            out: enabled=1
            out: gpgcheck=1
            out: gpgkey=https://download.docker.com/linux/centos/gpg
            out:
            out: [docker-ce-stable-source]
            out: name=Docker CE Stable - Sources
            out: baseurl=https://download.docker.com/linux/centos/$releasever/source/stable
            out: enabled=0
            out: gpgcheck=1
            out: gpgkey=https://download.docker.com/linux/centos/gpg
            out:
            out: [docker-ce-test]
            out: name=Docker CE Test - $basearch
            out: baseurl=https://download.docker.com/linux/centos/$releasever/$basearch/test
            out: enabled=0
            out: gpgcheck=1
            out: gpgkey=https://download.docker.com/linux/centos/gpg
            out:
            out: [docker-ce-test-source]
            out: name=Docker CE Test - Sources
            out: baseurl=https://download.docker.com/linux/centos/$releasever/source/test
            out: enabled=0
            out: gpgcheck=1
            out: gpgkey=https://download.docker.com/linux/centos/gpg
            out : https://download.docker.com/linux/centos/docker-ce.repo
            out  : docker-ce.repo
            out  : ParseResult(scheme='https', netloc='download.docker.com', path='/linux/centos/docker-ce.repo', params='', query='', fragment='')
            rpms : containerd.io-1.6.10-3.1.el9.aarch64.rpm
            rpms : containerd.io-1.6.11-3.1.el9.aarch64.rpm
            rpms : containerd.io-1.6.12-3.1.el9.aarch64.rpm
            rpms : containerd.io-1.6.13-3.1.el9.aarch64.rpm
            rpms : containerd.io-1.6.14-3.1.el9.aarch64.rpm
```

### CODE REVIEW

File: tmt/steps/prepare/artifact/providers/repository.py [hunk 1]
```diff
@@ -9,118 +10,219 @@
...
+    provides_artifact_provider,
 )
 from tmt.steps.prepare.artifact.providers.koji import RpmArtifactInfo
 from tmt.steps.provision import Guest
 from tmt.utils import GeneralError, Path, ShellScript
 
 
-@container
-class RepositoryFile:
+class RepositoryManager:
```
Comment (AthreyVinay) [inline tmt/steps/prepare/artifact/providers/repository.py]: Since all methods are @staticmethod and there’s no state, should these just be module-level functions instead of a class? Seems more pythonic.

File: tmt/steps/prepare/artifact/providers/repository.py [hunk 2]
```diff
@@ -9,118 +10,219 @@
...
-        # from the repository. Currently using an empty list as a temporary solution.
-        self._rpm_list = []
-
     @property
     def artifacts(self) -> Sequence[RpmArtifactInfo]:
-        raise NotImplementedError
+        """
+        List all RPMs discovered from the repository.
+
+        Note: The `fetch_contents` method must be called first to populate
```
Comment (AthreyVinay) [inline tmt/steps/prepare/artifact/providers/repository.py]: Just looking at the existing codebase, the `note` should be defined like so,
```rst
.. note:: The `fetch_contents` method must be called first to populate

File: tmt/steps/prepare/artifact/providers/repository.py [hunk 3]
```diff
@@ -9,118 +10,219 @@
...
     """
-    A helper class representing a repository .repo file from a URL.
+    A utility class for managing DNF repositories on a guest.
     """
 
-    url: str
-
-    def __post_init__(self) -> None:
+    @staticmethod
+    def enable_repository(
```
Comment (AthreyVinay) [inline tmt/steps/prepare/artifact/providers/repository.py]: Redundant parsing:
`enable_repository()` → returns repo_filepath
`get_repository_id() `→ parses repo file to get repo_id
`fetch_rpms() `→ parses repo file again to get the same repo_id

A possible solution is for `enable_repository()` to just return a `tuple[Path, str]` while we have the context maybe?

File: tmt/steps/prepare/artifact/providers/repository.py [hunk 4]
```diff
@@ -9,118 +10,219 @@
...
+
+        :param guest: The guest to operate on.
+        :param url: The URL of the .repo file to download.
+        :param repo_filename: The target filename for the .repo file.
+        :param logger: The logger for outputting messages.
+        :return: The path to the newly created repository file on the guest.
+        :raises DownloadError: If the download fails.
         """
+        repo_dest = Path("/etc/yum.repos.d") / repo_filename
+        sudo = "sudo " if not guest.facts.is_superuser else ""
```
Comment (therazix) [inline tmt/steps/prepare/artifact/providers/repository.py]: We can now use `guest.facts.sudo_prefix` instead.

File: tmt/steps/prepare/artifact/providers/repository.py [hunk 5]
```diff
@@ -9,118 +10,219 @@
...
+        """
+        List all RPMs discovered from the repository.
+
+        Note: The `fetch_contents` method must be called first to populate
+        the artifact list from the guest.
+        """
+        if self._rpm_list is None:
+            raise GeneralError(
+                "RPM list not available. The 'fetch_contents' method must be "
+                "called before 'list_artifacts'."
```
Comment (therazix) [inline tmt/steps/prepare/artifact/providers/repository.py]: There is no `list_artifacts` method.

Comment (AthreyVinay) [inline tmt/steps/prepare/artifact/providers/repository.py]: `list_artifacts` doesn't exist.

Comment (AthreyVinay) [reply tmt/steps/prepare/artifact/providers/repository.py]: POinted out here already: https://github.com/teemtee/tmt/pull/4139#discussion_r2419540067

File: tmt/steps/prepare/artifact/providers/repository.py (hunk 4 above)
Comment (vaibhavdaren) [reply tmt/steps/prepare/artifact/providers/repository.py]: addressed in https://github.com/teemtee/tmt/pull/4139/commits/c2d727be5cb44fb78dca33728604cdb025da04bd

File: tmt/steps/prepare/artifact/providers/repository.py (hunk 5 above)
Comment (vaibhavdaren) [reply tmt/steps/prepare/artifact/providers/repository.py]: addressed in https://github.com/teemtee/tmt/pull/4139/commits/c2d727be5cb44fb78dca33728604cdb025da04bd

File: tmt/steps/prepare/artifact/providers/repository.py [hunk 6]
```diff
@@ -9,118 +10,220 @@
...
-                raise ValueError
-        except ValueError as exc:
-            raise GeneralError(f"Invalid URL format for .repo file: '{self.url}'.") from exc
+            # Use -L to follow redirects and --fail to error out on HTTP errors.
+            script = ShellScript(
+                f"{guest.facts.sudo_prefix}curl -L --fail -o {quote(str(repo_dest))} {quote(url)}"
+            )
+            guest.execute(script, silent=True)
+        except GeneralError as error:
+            raise DownloadError(f"Failed to download repository file to '{repo_dest}'.") from error
```
Comment (AthreyVinay) [inline tmt/steps/prepare/artifact/providers/repository.py]: Nice to have more context here jsut for better debugging:
`raise DownloadError(f"Failed to download repository file from '{url}' to '{repo_dest}'.") from error`

File: tmt/steps/prepare/artifact/providers/repository.py [hunk 7]
```diff
@@ -9,118 +10,220 @@
...
+        repo_dest = Path("/etc/yum.repos.d") / repo_filename
         try:
-            self._parsed_url = urlparse(self.url)
-            if not self._parsed_url.scheme or not self._parsed_url.netloc:
-                raise ValueError
-        except ValueError as exc:
-            raise GeneralError(f"Invalid URL format for .repo file: '{self.url}'.") from exc
+            # Use -L to follow redirects and --fail to error out on HTTP errors.
+            script = ShellScript(
+                f"{guest.facts.sudo_prefix}curl -L --fail -o {quote(str(repo_dest))} {quote(url)}"
```
Comment (therazix) [inline tmt/steps/prepare/artifact/providers/repository.py]: `guest.facts.sudo_prefix` doesn’t include a trailing space.
```suggestion
                f"{guest.facts.sudo_prefix} curl -L --fail -o {quote(str(repo_dest))} {quote(url)}"
```

File: tmt/steps/prepare/artifact/providers/repository.py (hunk 1 above)
Comment (happz) [reply tmt/steps/prepare/artifact/providers/repository.py]: That's questionable. A class may bring a needed namespacing of a family of related functions.

File: tmt/steps/prepare/artifact/providers/repository.py [hunk 8]
```diff
@@ -9,118 +10,224 @@
...
+    provides_artifact_provider,
 )
 from tmt.steps.prepare.artifact.providers.koji import RpmArtifactInfo
 from tmt.steps.provision import Guest
 from tmt.utils import GeneralError, Path, ShellScript
 
 
-@container
-class RepositoryFile:
+class RepositoryManager:
```
Comment (happz) [inline tmt/steps/prepare/artifact/providers/repository.py]: IIUIC, `RepositoryManager` is a base class for doing stuff with a package repository. As such, I would move it into `__init__.py` as it's going to be used by other providers, or even by the plugin itself, and as a core piece of functionality it should not be provided by a plugin.

Comment (happz) [reply tmt/steps/prepare/artifact/providers/repository.py]: And maybe a slightly different approach: would it make sense to have `Repository` class, representing a repository? `Repository(name, filepath)`? With `enable()` method? I would keep `guest` and `logger` as inputs to `enable()`, at least for now, but the static nature of "manager" seems overly complicated.

File: tmt/steps/prepare/artifact/providers/repository.py (hunk 3 above)
Comment (happz) [reply tmt/steps/prepare/artifact/providers/repository.py]: To "enable" repository is something slightly different: a repository exists, it's known to dnf, it's just disabled. This method is more along the lines of "install repository".

File: tmt/steps/prepare/artifact/providers/repository.py [hunk 9]
```diff
@@ -9,118 +10,224 @@
...
+        repo_id = result.stdout.strip() if result.stdout else ""
+        if not repo_id:
+            raise GeneralError(f"Could not extract repository ID from '{repo_filepath}'.")
+
+        return repo_id
 
+    @staticmethod
+    def fetch_rpms(
+        guest: Guest, repo_filepath: Path, logger: tmt.log.Logger
+    ) -> list[RpmArtifactInfo]:
```
Comment (happz) [inline tmt/steps/prepare/artifact/providers/repository.py]: Aligning with the more property-like approach, this would be `rpms` property rather than an "action" like "fetch" - as a caller, I don't care about fetching, I just want packages. Guest and logger are a complication, and we will need to do something about them, because Koji providers do need them, but in different points of their workflow...

File: tmt/steps/prepare/artifact/providers/repository.py (hunk 1 above)
Comment (AthreyVinay) [reply tmt/steps/prepare/artifact/providers/repository.py]: Sure - but the module itself does serve as the namespace too.
eg: `repository.enable_repository()`

File: tmt/steps/prepare/artifact/providers/repository.py (hunk 2 above)
Comment (happz) [reply tmt/steps/prepare/artifact/providers/repository.py]: Actually:

```rst
.. note::

   The :py:meth:`fetch_contents` method must be called first to populate...
```

File: tmt/steps/prepare/artifact/providers/repository.py (hunk 1 above)
Comment (happz) [reply tmt/steps/prepare/artifact/providers/repository.py]: True, but that's where it hits https://github.com/teemtee/tmt/pull/4139#discussion_r2427176819 - it's "enable repo", "get repo id", "make repo dance" - all speak about a repository, which seems like there is some kind of "repository" object these actions should belong to...

Comment (AthreyVinay) [reply tmt/steps/prepare/artifact/providers/repository.py]: Agreed.

File: tests/unit/artifact/test_repository.py [hunk 10]
```diff
@@ -0,0 +1,216 @@
...
+    """Provides a MagicMock for the logger."""
+    return MagicMock()
+
+
+@pytest.fixture
+def mock_guest():
+    """Provides a MagicMock for the guest."""
+    guest = MagicMock()
+    guest.facts.is_superuser = False
+    # This is the fix: Explicitly set the sudo_prefix on the mock guest.
```
Comment (thrix) [inline tests/unit/artifact/test_repository.py]: How should I understand this comment?

File: tmt/steps/prepare/artifact/providers/repository.py [hunk 11]
```diff
@@ -9,118 +10,224 @@
...
+        repo_dest = Path("/etc/yum.repos.d") / repo_filename
         try:
-            self._parsed_url = urlparse(self.url)
-            if not self._parsed_url.scheme or not self._parsed_url.netloc:
-                raise ValueError
-        except ValueError as exc:
-            raise GeneralError(f"Invalid URL format for .repo file: '{self.url}'.") from exc
+            # Use -L to follow redirects and --fail to error out on HTTP errors.
+            script = ShellScript(
+                f"{guest.facts.sudo_prefix} curl -L --fail -o {quote(str(repo_dest))} {quote(url)}"
```
Comment (thrix) [inline tmt/steps/prepare/artifact/providers/repository.py]: We should start with reasonable retries from day one for these curl commands

File: tmt/steps/prepare/artifact/providers/repository.py [hunk 12]
```diff
@@ -9,118 +10,224 @@
...
-    def __str__(self) -> str:
-        return self.filename
+        :param guest: The guest to query.
+        :param repo_filepath: The path to the .repo file on the guest.
+        :param logger: The logger for outputting messages.
+        :return: The repository ID.
+        :raises GeneralError: If the repo ID cannot be extracted.
+        """
+        script = ShellScript(
+            f"grep '^\\[' {quote(str(repo_filepath))} | head -n 1 | sed 's/^\\[\\(.*\\)\\]$/\\1/'"
```
Comment (thrix) [inline tmt/steps/prepare/artifact/providers/repository.py]: Note that a repo file can contain multiple repositories. If it is a decision that we will support only one, please mention it.

Also, I am wondering if it would not be cleaner to read the repository file content and parse it via `ini` parser rather? We are in this situation in Testing Farm that a lot of stuff is done via commands and the guest, and it is becoming harder and harder to maintain. Maybe rather reading the data from the guest and parsing it nicely in python would be a better approach?

File: tmt/steps/prepare/artifact/providers/repository.py [hunk 13]
```diff
@@ -9,118 +10,224 @@
...
+        :raises GeneralError: If dnf query fails.
+        """
+        # 1. Get the repository ID.
+        repo_id = RepositoryManager.get_repository_id(guest, repo_filepath, logger)
+
+        # 2. List all available packages from that repository using a robust query format.
+        # We query for name, epoch, version, release, and architecture.
+        qf = "'%{name} %{epoch} %{version} %{release} %{arch}'"
+        script = ShellScript(
+            "dnf repoquery --refresh --disablerepo='*' "
```
Comment (thrix) [inline tmt/steps/prepare/artifact/providers/repository.py]: Please file an issue so we do not forget about `yum` support here, this will not work on RHEL6 and RHEL7 at least.

File: tests/unit/artifact/test_repository.py [hunk 14]
```diff
@@ -0,0 +1,216 @@
...
+        # Verify first call (get_repository_id)
+        command = (
+            f"grep '^\\[' {quote(str(repo_filepath))} | head -n 1 | sed 's/^\\[\\(.*\\)\\]$/\\1/'"
+        )
+        assert str(calls[0][0][0]) == command
+
+        # Verify second call (dnf repoquery)
+        command = (
+            "dnf repoquery --refresh --disablerepo='*' "
+            f"--enablerepo={quote(repo_id)} --available "
```
Comment (thrix) [inline tests/unit/artifact/test_repository.py]: According to dnf documentation the `--available` is the default?

```
       --available
              Query available packages.
              This is the default behavior.
              Can be combined with --installed to query both installed and available packages.
 ```

File: tmt/steps/prepare/artifact/providers/repository.py (hunk 12 above)
Comment (vaibhavdaren) [reply tmt/steps/prepare/artifact/providers/repository.py]: [1] I will be adding support for `multiple` repositories in a repository_file. Thanks for pointing this out. ( I totally missed this) 
[2] Nice suggestion. I will be changing the approach to use config parser as its much cleaner.

File: tmt/steps/prepare/artifact/providers/repository.py (hunk 13 above)
Comment (vaibhavdaren) [reply tmt/steps/prepare/artifact/providers/repository.py]: It makes more sense to  create and extend `guest.package_manager.repoquery`.
I will be adding a FIXME note and file an issue for extending support for various package managers.

File: tmt/steps/prepare/artifact/providers/repository.py (hunk 11 above)
Comment (vaibhavdaren) [reply tmt/steps/prepare/artifact/providers/repository.py]: Added retry logic in the next commit.

File: tmt/steps/prepare/artifact/providers/repository.py (hunk 8 above)
Comment (vaibhavdaren) [reply tmt/steps/prepare/artifact/providers/repository.py]: I have changed the structure of class according to your feedback. 
We create a Repository Object with just the `url` and `file_name`.
We pass Logger and Guest , when we want to install the repository.

File: tmt/steps/prepare/artifact/providers/repository.py (hunk 9 above)
Comment (vaibhavdaren) [reply tmt/steps/prepare/artifact/providers/repository.py]: I have changed rpms to be a property of `Repository` class. the only catch when we convert this to property is that `guest` and `logger` params are now implicitly  set during the `install` method.

File: tmt/steps/prepare/artifact/providers/__init__.py [hunk 15]
```diff
@@ -195,6 +202,174 @@ def _filter_artifacts(self, exclude_patterns: list[Pattern[str]]) -> Iterator[Ar
...
+        :raises GeneralError: If the .repo file is empty, not readable, cannot be parsed
+                             as INI, or contains no repository sections.
+        """
+        self.logger = logger
+        self.guest = guest
+        self.filepath = Path("/etc/yum.repos.d") / self.filename
+        script = ShellScript(
+            f"{guest.facts.sudo_prefix} curl -L --fail -o {quote(str(self.filepath))} "
+            f"{quote(self.url)}"
+        )
```
Comment (tcornell-bus) [inline tmt/steps/prepare/artifact/providers/__init__.py]: I think it was voted in the tmt hacking session a while ago (2025-05-20) to use a slightly different format:
```
rf"""
...
...
"""
```
instead of:
```
f"..."
f"..."
```
It was to be documented, but I do not know if the change to docs ever happened.

This is a short script though, so likely not an issue.

**Comment by vaibhavdaren** [comment]
FYI, AS per discussion on the review meeting the big PR is split into multiple chunks. 
I've put together a few PRs to polish up the package manager and repository support. Here's what they cover
https://github.com/teemtee/tmt/pull/4206: This adds repository handling directly into the package manager.
https://github.com/teemtee/tmt/pull/4205: This introduces a new class to manage repository-related functionality.
https://github.com/teemtee/tmt/pull/4207: This refactors the repository provider to work with the new package manager and RepositoryClass.

**Review by AthreyVinay — APPROVED** [review]
LGTM :)

File: tmt/steps/prepare/artifact/providers/__init__.py [hunk 16]
```diff
@@ -195,6 +202,174 @@ def _filter_artifacts(self, exclude_patterns: list[Pattern[str]]) -> Iterator[Ar
...
+                f"Failed to download repository file from '{self.url}' to "
+                f"'{self.filepath}': {last_error!s}"
+            ) from last_error
+
+        try:
+            result = guest.execute(ShellScript(f'{guest.facts.sudo_prefix} cat {self.filepath}'))
+            content = result.stdout
+
+            if content is None:
+                raise GeneralError("cat command did not return any output.")
```
Comment (tcornell-bus) [inline tmt/steps/prepare/artifact/providers/__init__.py]: It would be nice to know what file `cat` was run on.

**Review by tcornell-bus — APPROVED** [review]
LGTM

File: tmt/steps/prepare/artifact/providers/repository.py [hunk 17]
```diff
@@ -1,126 +1,188 @@
...
+                # Catch any other unexpected errors
+                self.logger.debug(f"Unexpected error while parsing package '{pkg}': {error}.")
+                continue
+
+        self.logger.debug(f"Successfully discovered '{len(self._artifact_list)}' artifacts.")
+
         return []
+
+
+# FIXME: Make this function more robust. The current regex-based parsing
```
Comment (happz) [inline tmt/steps/prepare/artifact/providers/repository.py:136]: Please, mention it in https://github.com/teemtee/tmt/issues/4221

File: tmt/steps/prepare/artifact/providers/repository.py [hunk 18]
```diff
@@ -1,126 +1,188 @@
...
+    # 4: Version (must contain at least one digit)
+    r"(?P<version>[^-:]*\d[^-:]*)"
+    r"-"  # (literal hyphen)
+    r"(?P<release>[^-]+)"  # 5: Release (no hyphens)
+    r"\."  # (literal dot)
+    r"(?P<arch>[^.]+)$"  # 6: Arch
+)
+
+
+def parse_rpm_string(pkg_string: str) -> dict[str, str]:
```
Comment (happz) [inline tmt/steps/prepare/artifact/providers/repository.py:160]: Another point for https://github.com/teemtee/tmt/issues/4221 - investigate how to make the "rpm" more hidden, as it's just one kind of repositories out there. We may end up with some `RpmRepository`, or `YumRepository` implementations of a general `Repository` abstract class later. Let's just track it for now.

File: tmt/steps/prepare/artifact/providers/repository.py [hunk 19]
```diff
@@ -1,126 +1,188 @@
...
+        List all RPMs discovered from the repositories.
+
+        .. note::
+
+            The :py:meth:`fetch_contents` method must be called first to populate
+            the artifact list from the guest.
+        """
+        # Check for None to see if fetch_contents() has been called
+        if self._artifact_list is None:
+            raise tmt.utils.GeneralError("Call fetch_contents first to discover artifacts.")
```
Comment (happz) [inline tmt/steps/prepare/artifact/providers/repository.py:73]: Another point for https://github.com/teemtee/tmt/issues/4221: we need to make sure to catch this ordering in a different way. A runtime exception is too late, we need some type-based or workflow-based guarantees. Good for now, but let's prevent the ordering mistake later.

Comment (vaibhavdaren) [reply tmt/steps/prepare/artifact/providers/repository.py:73]: Added.

File: tmt/steps/prepare/artifact/providers/repository.py (hunk 17 above)
Comment (vaibhavdaren) [reply tmt/steps/prepare/artifact/providers/repository.py:136]: Added.

File: tmt/steps/prepare/artifact/providers/repository.py (hunk 18 above)
Comment (vaibhavdaren) [reply tmt/steps/prepare/artifact/providers/repository.py:160]: Added

**Review by LecrisUT — APPROVED** [review]

**Comment by vaibhavdaren** [comment]
@happz : Added improvements to checklist issue : https://github.com/teemtee/tmt/issues/4221.

File: tmt/steps/prepare/artifact/providers/repository.py (hunk 17 above)
Comment (thrix) [reply tmt/steps/prepare/artifact/providers/repository.py:136]: @vaibhavdaren with that this should be covered well with unit tests, we could for example list all rpms in fedora repositories now and make sure it works, or go ahead with some generation like hypothesis to generate the possible nvrs to sort out all the corner cases

File: tmt/steps/prepare/artifact/providers/repository.py [hunk 20]
```diff
@@ -1,126 +1,188 @@
...
+
+
+# FIXME: Make this function more robust. The current regex-based parsing
+# is a "happy path" implementation and will fail on complex or
+# unusually-named packages. This is acceptable for now but should
+# be hardened later
+# Regex to parse N-E:V-R.A format.
+# Groups: 1:Name, 3:Epoch (optional), 4:Version, 5:Release, 6:Arch
+_PKG_REGEX = re.compile(
+    r"^(?P<name>.+?)"  # 1: Name (non-greedy)
```
Comment (thrix) [inline tmt/steps/prepare/artifact/providers/repository.py]: What is the purpose of non-greedy here? it seems to be it is not needed?

File: tmt/steps/prepare/artifact/providers/repository.py [hunk 21]
```diff
@@ -1,126 +1,188 @@
...
+    r"-"  # (literal hyphen)
+    r"((?P<epoch>\d+):)?"  # 2, 3: Epoch (optional group)
+    # 4: Version (must contain at least one digit)
+    r"(?P<version>[^-:]*\d[^-:]*)"
+    r"-"  # (literal hyphen)
+    r"(?P<release>[^-]+)"  # 5: Release (no hyphens)
+    r"\."  # (literal dot)
+    r"(?P<arch>[^.]+)$"  # 6: Arch
```
Comment (thrix) [inline tmt/steps/prepare/artifact/providers/repository.py]: you can drop the ending $ and use `re.fullmatch` instead

File: tmt/steps/prepare/artifact/providers/repository.py [hunk 22]
```diff
@@ -1,126 +1,188 @@
...
+    :param pkg_string: The package string, e.g., "docker-ce-1:20.10.7-3.el8.x86_64".
+    :raises ValueError: if the package string is malformed.
+    :return: A dictionary of RPM components.
+    """
+
+    # 1. Match the package string against the regex
+    match = _PKG_REGEX.match(pkg_string)
+
+    if not match:
+        raise ValueError(f"String '{pkg_string}' does not match N-V-R.A format")
```
Comment (thrix) [inline tmt/steps/prepare/artifact/providers/repository.py]:
```suggestion
        raise ValueError(f"String '{pkg_string}' does not match N-E:V-R.A format")
```

File: tmt/steps/prepare/artifact/providers/repository.py [hunk 23]
```diff
@@ -1,126 +1,188 @@
...
+    release = match.group('release')
+    arch = match.group('arch')
+
+    # Optional epoch group can be None
+    epoch = match.group('epoch')
+    if epoch is None:
+        epoch = '0'
+
+    # Additional validation
+    # RPM package names are not allowed to contain colons.
```
Comment (thrix) [inline tmt/steps/prepare/artifact/providers/repository.py]: can we add more info on when this can happen and why this additional validation is needed pls?

File: tmt/steps/prepare/artifact/providers/repository.py (hunk 18 above)
Comment (thrix) [reply tmt/steps/prepare/artifact/providers/repository.py:160]: Also maybe it could return some better data type then a dict in the future?

Comment (thrix) [inline tmt/steps/prepare/artifact/providers/repository.py:160]: @vaibhavdaren @happz one stupid question, what if the repository contains also source rpms (.src.rpm) are we somehow dealing with that somewhere?

File: tmt/steps/prepare/artifact/providers/repository.py [hunk 24]
```diff
@@ -1,126 +1,188 @@
...
         return []
+
+
+# FIXME: Make this function more robust. The current regex-based parsing
+# is a "happy path" implementation and will fail on complex or
+# unusually-named packages. This is acceptable for now but should
+# be hardened later
+# Regex to parse N-E:V-R.A format.
+# Groups: 1:Name, 3:Epoch (optional), 4:Version, 5:Release, 6:Arch
+_PKG_REGEX = re.compile(
```
Comment (happz) [inline tmt/steps/prepare/artifact/providers/repository.py:142]: Check the `tmt.hardware` package for its use of verbose patterns, `re.VERBOSE`. No need for `r` on every line, comments, fully supported, much nicer to read.

File: tmt/steps/prepare/artifact/providers/repository.py (hunk 20 above)
Comment (vaibhavdaren) [reply tmt/steps/prepare/artifact/providers/repository.py]: Moved to greedy in the next commit.

File: tmt/steps/prepare/artifact/providers/repository.py (hunk 21 above)
Comment (vaibhavdaren) [reply tmt/steps/prepare/artifact/providers/repository.py]: Thanks. Used full match in the next commit.

File: tmt/steps/prepare/artifact/providers/repository.py (hunk 18 above)
Comment (vaibhavdaren) [reply tmt/steps/prepare/artifact/providers/repository.py:160]: Nice catch, I have filtered them out specifically in the next commit.

Comment (vaibhavdaren) [reply tmt/steps/prepare/artifact/providers/repository.py:160]: > I think a better way to be return some kind of package dataclass from list_packages. 

We could refactor list_packages to return a list of a specific dataclass (e.g., PackageInfo).

The primary benefit would be removing the need for all the regex parsing. Instead, we could leverage repoquery's queryformat option to get the exact data we need. The output could then be parsed simply (perhaps with just string.split(' ')) to instantiate the dataclass objects directly.

File: tmt/steps/prepare/artifact/providers/repository.py (hunk 23 above)
Comment (vaibhavdaren) [reply tmt/steps/prepare/artifact/providers/repository.py]: only malformed package names.

File: tmt/steps/prepare/artifact/providers/repository.py (hunk 24 above)
Comment (vaibhavdaren) [reply tmt/steps/prepare/artifact/providers/repository.py:142]: Thanks for the tip. Indeed that style is very clean.

File: tmt/steps/prepare/artifact/providers/repository.py [hunk 25]
```diff
@@ -1,126 +1,195 @@
...
+
+            except ValueError:
+                # Catches both regex failing to match (ValueError)
+                # or an explicit ValueError raised by the utility function.
+                self.logger.warning(f"Failed to parse malformed package string '{pkg}'. Skipping.")
+                continue
+
+            except Exception as error:
+                # Catch any other unexpected errors
+                self.logger.debug(f"Unexpected error while parsing package '{pkg}': {error}.")
```
Comment (happz) [inline tmt/steps/prepare/artifact/providers/repository.py]: `tmt.utils.show_exception_as_warning()`, please.

File: tmt/steps/prepare/artifact/providers/repository.py (hunk 18 above)
Comment (thrix) [reply tmt/steps/prepare/artifact/providers/repository.py:160]: @vaibhavdaren I would add it to #4221

Comment (thrix) [reply tmt/steps/prepare/artifact/providers/repository.py:160]: or another follow up issue if not suitable

File: tmt/steps/prepare/artifact/providers/repository.py (hunk 20 above)
Comment (thrix) [reply tmt/steps/prepare/artifact/providers/repository.py]: thank you

File: tmt/steps/prepare/artifact/providers/repository.py (hunk 23 above)
Comment (thrix) [reply tmt/steps/prepare/artifact/providers/repository.py]: ack

File: tmt/steps/prepare/artifact/providers/repository.py (hunk 18 above)
Comment (thrix) [reply tmt/steps/prepare/artifact/providers/repository.py:160]: Ack, we might need it to be available though, so maybe another follow up issue to resolve? I believe we do not want it parse here of course, but it should be downloaded and available possibly?

File: tmt/steps/prepare/artifact/providers/repository.py (hunk 20 above)
Comment (LecrisUT) [reply tmt/steps/prepare/artifact/providers/repository.py]: For some reason I thought this was the same PR as #4207, but the regex there [^1] is basically identical and should be de-duplicated. But my view on this regex, if it is simple enough, you do not need to document it. My suggestion is to go with https://regex101.com/r/49HQNy/1 which works because of the greedy matches.

In the current iteration I do not like the unnecessary exclusion matching. It is technically correct and equivalent, but it makes it harder to parse and read. And if you wish to document it, then something like a regex101 (or any equivalent) would be much better because it can have relevant examples included, and it has a more navigable breakdown of the logic (even though the presentation there is also way too verbose and technical)

Anyway not a required change for this PR, but just to keep it in mind

[^1]: https://github.com/teemtee/tmt/pull/4207#discussion_r2478772300

File: tmt/steps/prepare/artifact/providers/repository.py (hunk 25 above)
Comment (vaibhavdaren) [reply tmt/steps/prepare/artifact/providers/repository.py]: Thankyou for suggestion added in next commit. I think I will go through the utility files today. It has some very powerful tools.

File: tmt/steps/prepare/artifact/providers/repository.py (hunk 18 above)
Comment (vaibhavdaren) [reply tmt/steps/prepare/artifact/providers/repository.py:160]: Added a `TODO`: and issue so we dont forget about it : https://github.com/teemtee/tmt/issues/4293 
In the review meeting today we decided that we need them in some cases like dist git. Keeping a note.

Comment (vaibhavdaren) [reply tmt/steps/prepare/artifact/providers/repository.py:160]: Added.

Comment (thrix) [reply tmt/steps/prepare/artifact/providers/repository.py:160]: thank you!

File: tmt/steps/prepare/artifact/providers/repository.py (hunk 21 above)
Comment (psss) [reply tmt/steps/prepare/artifact/providers/repository.py]: Which commit? I don't see `fullmatch` in the change?

Comment (vaibhavdaren) [reply tmt/steps/prepare/artifact/providers/repository.py]: Added in https://github.com/teemtee/tmt/pull/4139/commits/beea208ac5e40fcd54ce4d0b5a0f4ece712446c2

File: tmt/steps/prepare/artifact/providers/repository.py (hunk 20 above)
Comment (happz) [reply tmt/steps/prepare/artifact/providers/repository.py]: What is the outcome of this thread?

Comment (LecrisUT) [reply tmt/steps/prepare/artifact/providers/repository.py]: Covered in #4221

> Make the regex-based parsing function more robust which is used to convert the PackageNames to RPM objects.

I will add the link to my comment there also

**Comment by psss** [comment]
I believe all conversations should now be covered. Merging.
